errors/error.go

```diff
@@ -1,94 +1,237 @@
+		Detail:   e.Detail,
+		Status:   strconv.Itoa(e.Status),
+	})
+}
 
-const (
-	// PutErrorNil indicates that no error occurred during handling a PUT HTTP request.
-	PutErrorNil PutError = iota
-	// PutErrorUniqueness shall be returned when one or more of the attribute values are already in use or are reserved.
-	PutErrorUniqueness
-	// PutErrorMutability shall be returned when the attempted modification is not compatible with the target
-	// attribute's mutability or current state.
-	PutErrorMutability
-	// PutErrorResourceNotFound returns an error with status code 404 and a human readable message containing the
-	// identifier of the resource that was requested to be replaced but not found.
-	PutErrorResourceNotFound
-	// PutErrorNotImplemented allows consumers to create a get handler that simply returns an unsupported error.
-	PutErrorNotImplemented
-)
+// UnmarshalJSON converts the error json data to its corresponding struct representation.
+func (e *ScimError) UnmarshalJSON(data []byte) error {
+	var tmpScimError struct {
+		ScimType ScimType
+		Detail   string
+		Status   string
+	}
 
-// DeleteError represents an error that is returned by a DELETE HTTP request.
-type DeleteError int
+	err := json.Unmarshal(data, &tmpScimError)
+	if err != nil {
+		return err
+	}
 
-const (
-	// DeleteErrorNil indicates that no error occurred during handling a DELETE HTTP request.
-	DeleteErrorNil DeleteError = iota
-	// DeleteErrorResourceNotFound returns an error with status code 404 and a human readable message containing the
-	// identifier of the resource that was requested to be deleted but not found.
-	DeleteErrorResourceNotFound
-	// DeleteErrorNotImplemented allows consumers to create a get handler that simply returns an unsupported error.
-	DeleteErrorNotImplemented
+	status, err := strconv.Atoi(tmpScimError.Status)
+	if err != nil {
+		return err
+	}
+
+	*e = ScimError{
+		ScimType: tmpScimError.ScimType,
+		Detail:   tmpScimError.Detail,
+		Status:   status,
+	}
+
+	return nil
+}
+
+// ScimErrorResourceNotFound returns an 404 SCIM error with a detailed message based on the id.
+func ScimErrorResourceNotFound(id string) ScimError {
+	return ScimError{
+		Detail: fmt.Sprintf("Resource %s not found.", id),
+		Status: http.StatusNotFound,
+	}
+}
+
+// ScimErrorBadParams returns an 400 SCIM error with a detailed message based on the invalid parameters.
+func ScimErrorBadParams(invalidParams []string) ScimError {
+	var suffix string
+
+	if len(invalidParams) > 1 {
+		suffix = "s"
+	}
+
+	return ScimErrorBadRequest(fmt.Sprintf(
+		"Bad Request. Invalid parameter%s provided in request: %s.",
+		suffix,
+		strings.Join(invalidParams, ", "),
+	))
+}
+
+// ScimErrorBadRequest returns an 400 SCIM error with the given message.
+func ScimErrorBadRequest(msg string) ScimError {
+	return ScimError{
+		Detail: msg,
+		Status: http.StatusBadRequest,
+	}
+}
+
+var (
+	// ScimErrorUniqueness returns an 409 SCIM error with a detailed message.
```

Comment:
You can create external vars for all of the error types
```suggestion
	// ScimErrorInvalidFilter returns an 409 SCIM error with a detailed message.
	ScimErrorInvalidFilter = ScimError{
		ScimType: ScimTypeInvalidFilter,
		Detail:   "The specified filter syntax was invalid (does not comply with Figure 1), or the specified attribute and filter comparison combination is not supported.",
		Status:   http.StatusBadRequest,
	}
	// ScimErrorTooMany returns an 409 SCIM error with a detailed message.
	ScimErrorTooMany = ScimError{
		ScimType: ScimTypeTooMany,
		Detail:   "The specified filter yields many more results than the server is willing to calculate or process.  For example, a filter such as \"(userName pr)\" by itself would return all entries with a \"userName\" and MAY not be acceptable to the service provider.",
		Status:   http.StatusBadRequest,
	}
	// ScimErrorUniqueness returns an 409 SCIM error with a detailed message.
```